// json-obj-form-generator/dist/components/designer/ErrorContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _icons = require("./icons");

var _propTypes2 = require("./propTypes");

var _utils = require("./utils");

var _InputWithTip = _interopRequireDefault(require("./InputWithTip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ErrorContainer = /*#__PURE__*/function (_Component) {
  _inherits(ErrorContainer, _Component);

  var _super = _createSuper(ErrorContainer);

  function ErrorContainer() {
    var _this;

    _classCallCheck(this, ErrorContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isOpenErrors: false,
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "err-" + _this.props.jkey.sufix)
    });

    return _this;
  }

  _createClass(ErrorContainer, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var filled = true;
      var keys = Object.keys(this.props.values);
      var show = keys.length > 0;

      for (var i = 0; i < keys.length; i++) {
        if (this.props.errors[keys[i]] === undefined) {
          show = false;
          break;
        } else {
          if (this.props.values[keys[i]].length === 0) {
            filled = false;
          }
        }
      }

      if (show) {
        return /*#__PURE__*/_react.default.createElement(_reactstrap.CardFooter, {
          className: "jofgen-D-error-collapse" + (!filled ? "-alert " : " ") + this.props.className,
          style: this.props.style
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardTitle, {
          style: {
            marginBottom: this.state.isOpenErrors ? undefined : "0"
          }
        }, /*#__PURE__*/_react.default.createElement("table", {
          style: {
            width: "100%"
          }
        }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", {
          className: "jofgen-D-children"
        }, filled ? this.props.icons.icon : /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
          id: this.state.gId + "-alert-errors",
          style: {
            display: "initial"
          },
          onMouseOver: function onMouseOver() {
            _this2.setState({
              infoShow: true
            });
          },
          onMouseOut: function onMouseOut() {
            _this2.setState({
              infoShow: false
            });
          }
        }, this.props.icons.invalid_icon), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
          target: this.state.gId + "-alert-errors",
          isOpen: this.state.infoShow
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.translation.alert))), this.props.translation.title), /*#__PURE__*/_react.default.createElement("td", {
          style: {
            width: "50px",
            textAlign: "right"
          }
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
          size: this.props.size,
          color: "link",
          onClick: function onClick() {
            _this2.setState({
              isOpenErrors: !_this2.state.isOpenErrors
            });
          }
        }, /*#__PURE__*/_react.default.createElement("svg", {
          className: "jofgen-D-collapse-icon",
          style: {
            transform: this.state.isOpenErrors ? "rotate(180deg)" : "rotate(0deg)"
          },
          viewBox: "0 0 24 24"
        }, /*#__PURE__*/_react.default.createElement("path", {
          fill: "#000",
          d: "M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
        })))))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Collapse, {
          isOpen: this.state.isOpenErrors
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true,
          className: "mb-0"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: 12
        }, Object.keys(this.props.values).map(function (item, idx) {
          return /*#__PURE__*/_react.default.createElement(_InputWithTip.default, {
            key: idx,
            jkey: {
              prefix: _this2.state.gId,
              sufix: "-" + item
            },
            size: _this2.props.size,
            type: "textarea",
            title: _this2.props.errors[item].name,
            desc: _this2.props.errors[item].tip,
            value: _this2.props.values[item],
            placeholder: _this2.props.errors[item].placeHolder,
            onChange: function onChange(e) {
              _this2.props.onChange(item, e);
            }
          });
        })))));
      }

      return null;
    }
  }]);

  return ErrorContainer;
}(_react.Component);

exports.default = ErrorContainer;
ErrorContainer.propTypes = {
  errors: _propTypes.default.object,
  values: _propTypes.default.object,
  onChange: _propTypes.default.func.isRequired,
  translation: _propTypes.default.shape({
    title: _propTypes.default.string.isRequired,
    alert: _propTypes.default.string.isRequired
  }),
  icons: _propTypes.default.shape({
    icon: _propTypes.default.any.isRequired,
    invalid_icon: _propTypes.default.any.isRequired
  }),
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string
};
ErrorContainer.defaultProps = {
  jkey: _propTypes2.Default_keyPropType,
  className: "",
  size: "sm",
  errors: [],
  values: [],
  icons: {
    icon: _icons.error,
    invalid_icon: _icons.errorAlert
  },
  translation: {
    title: "Errors",
    errors: "Some errors are not defined"
  }
};
