// json-obj-form-generator/dist/components/designer/InputCustom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rebuild = exports.prototype = exports.valid = exports.clean = exports.InputCustom = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _propTypes2 = require("./propTypes");

var _utils = require("./utils");

var _icons = require("./icons");

var _ErrorContainer = _interopRequireDefault(require("./ErrorContainer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var InputCustom = /*#__PURE__*/function (_Component) {
  _inherits(InputCustom, _Component);

  var _super = _createSuper(InputCustom);

  function InputCustom() {
    var _this;

    _classCallCheck(this, InputCustom);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isOpenMain: true,
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "regx" + _this.props.jkey.sufix)
    });

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.value;
      props.uid = props.uid.replace(" ", "_");
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeBool", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.checked;
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeError", function (key, value) {
      var props = this.props.value;
      props[key] = value;
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "collapseMain", function () {
      this.setState({
        isOpenMain: !this.state.isOpenMain
      });
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(InputCustom, [{
    key: "renderErrors",
    value: function renderErrors(value, regex, midfix) {
      var errors = [];

      if (value !== undefined && value.length > 0) {
        try {
          if (regex.length > 0 && !new RegExp(regex).test(value)) {
            errors = [].concat(_toConsumableArray(errors), [/*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
              key: this.props.gId + "-" + midfix + "-cb-T",
              valid: false
            }, this.props.translation.messages["type"])]);
          }
        } catch (_unused) {
          errors = [].concat(_toConsumableArray(errors), [/*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
            key: this.props.gId + "-" + midfix + "-cb-T",
            valid: false
          }, this.props.translation.messages["rgxv"])]);
        }
      }

      return errors;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var invalid = !valid(this.props.value);
      var renderErrDefault = this.props.value.default.length > 0 ? this.renderErrors(this.props.value.default, this.props.value.regex, "d") : [];
      var renderErrValue = this.props.value.value.length > 0 ? this.renderErrors(this.props.value.value, this.props.value.regex, "v") : [];
      var renderErrRegex = [];

      try {
        var rgx = new RegExp(this.props.value.regex);
      } catch (_unused2) {
        renderErrRegex = [].concat(_toConsumableArray(renderErrRegex), [/*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          key: this.props.gId + "-RT-cb-RGX",
          valid: false
        }, this.props.translation.messages["rgx"])]);
      }

      var default_section;

      if (this.props.value.required) {
        default_section = /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: 6
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
          size: this.props.size,
          sm: 2,
          className: "jofgen-D-inputLabel"
        }, this.props.translation.default), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          className: "jofgen-D-input-col"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
          name: "default",
          type: "text",
          value: this.props.value.default,
          onChange: this.onChange,
          invalid: renderErrDefault.length > 0,
          bsSize: this.props.size
        }), renderErrDefault)));
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        className: this.props.className + " jofgen-D-card jofgen-D-regx" + (this.props.value.required ? " required" : "") + (invalid || this.props.invalid ? " invalid" : ""),
        style: this.props.style
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardBody, {
        className: "jofgen-D-card-body"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardTitle, null, /*#__PURE__*/_react.default.createElement("table", {
        style: {
          width: "100%"
        }
      }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", {
        className: "jofgen-D-children"
      }, invalid || this.props.invalid ? this.props.icons.invalid_icon : this.props.icons.icon, /*#__PURE__*/_react.default.createElement("b", null, this.props.translation.title)), /*#__PURE__*/_react.default.createElement("td", {
        style: {
          width: "50px",
          textAlign: "right"
        }
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
        size: this.props.size,
        color: "link",
        onClick: this.collapseMain
      }, /*#__PURE__*/_react.default.createElement("svg", {
        className: "jofgen-D-collapse-icon",
        style: {
          transform: this.state.isOpenMain ? "rotate(180deg)" : "rotate(0deg)"
        },
        viewBox: "0 0 24 24"
      }, /*#__PURE__*/_react.default.createElement("path", {
        fill: "#000",
        d: "M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      })))))))), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 2,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.uid), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "uid",
        type: "text",
        value: this.props.value.uid,
        onChange: this.onChange,
        invalid: this.props.value.uid.length === 0 || this.props.invalid,
        bsSize: this.props.size
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.uid), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        check: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        check: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "required",
        type: "checkbox",
        checked: this.props.value.required,
        onChange: this.onChangeBool
      }), " " + this.props.translation.required), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        check: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "is_password",
        type: "checkbox",
        checked: this.props.value.is_password,
        onChange: this.onChangeBool
      }), " " + this.props.translation.is_password))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: "6"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 2,
        className: "jofgen-D-inputLabel jofgen-D-inputLabelWithpopUp"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        className: "jofgen-D-col-form-label-sm",
        size: this.props.size
      }, this.props.translations.width), ["1", "2", "3", "4", "5"].includes(this.props.value.sm) ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        id: this.state.gId + "popup",
        style: {
          float: "right"
        },
        onMouseOver: function onMouseOver() {
          _this2.setState({
            alertShow: true
          });
        },
        onMouseOut: function onMouseOut() {
          _this2.setState({
            alertShow: false
          });
        }
      }, this.props.icons_set.alert), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
        target: this.state.gId + "popup",
        isOpen: this.state.alertShow
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.translations.smallWidthAlert))) : null), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CustomInput, {
        name: "sm",
        type: "select",
        value: this.props.value.sm,
        onChange: this.onChange,
        id: this.state.gId + "dropdown",
        bsSize: this.props.size
      }, /*#__PURE__*/_react.default.createElement("option", {
        value: 1
      }, "1 (12 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 2
      }, "2 (6 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 3
      }, "3 (4 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 4
      }, "4 (3 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 5
      }, "5"), /*#__PURE__*/_react.default.createElement("option", {
        value: 6
      }, "6 (2 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 7
      }, "7"), /*#__PURE__*/_react.default.createElement("option", {
        value: 8
      }, "8"), /*#__PURE__*/_react.default.createElement("option", {
        value: 9
      }, "9"), /*#__PURE__*/_react.default.createElement("option", {
        value: 10
      }, "10"), /*#__PURE__*/_react.default.createElement("option", {
        value: 11
      }, "11"), /*#__PURE__*/_react.default.createElement("option", {
        value: 12
      }, "12 (1 ", this.props.translations.columns, ")"))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.name), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "name",
        type: "text",
        value: this.props.value.name,
        onChange: this.onChange,
        bsSize: this.props.size
      })))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.placeholder), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "placeholder",
        type: "text",
        value: this.props.value.placeholder,
        onChange: this.onChange,
        bsSize: this.props.size
      })))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 12
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.tip), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "tip",
        type: "textarea",
        value: this.props.value.tip,
        onChange: this.onChange,
        bsSize: this.props.size
      }))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Collapse, {
        isOpen: this.state.isOpenMain
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 2,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.value), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "value",
        type: "text",
        value: this.props.value.value,
        onChange: this.onChange,
        invalid: renderErrValue.length > 0,
        bsSize: this.props.size
      }), renderErrValue))), default_section, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: this.props.value.required ? 12 : 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.regex), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "regex",
        type: "text",
        value: this.props.value.regex,
        onChange: this.onChange,
        invalid: renderErrRegex.length > 0,
        bsSize: this.props.size
      }), renderErrRegex, this.props.value.regex.length === 0 ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
        color: "info"
      }, this.props.translation.messages["norgx"]) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null))))))), /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_ErrorContainer.default, {
        jkey: {
          prefix: this.state.gId + "-err",
          sufix: ""
        },
        size: this.props.size,
        values: Object.assign({}, this.props.value.required ? {
          err_req: this.props.value.err_req
        } : null, this.props.value.regex.length > 0 && renderErrRegex.length === 0 ? {
          err_type: this.props.value.err_type
        } : null),
        errors: {
          err_req: this.props.translation.errors.err_req,
          err_type: this.props.translation.errors.err_type
        },
        translation: {
          title: this.props.translation.errors.title,
          alert: this.props.translation.errors.alert
        },
        onChange: this.onChangeError
      })), /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_ErrorContainer.default, {
        className: "jofgen-D-warning",
        size: this.props.size,
        jkey: {
          prefix: this.state.gId + "-",
          sufix: "warn"
        },
        icons: {
          icon: this.props.icons.warnings,
          invalid_icon: this.props.icons.warningsAlert
        },
        values: Object.assign({}, this.props.value.default.length > 0 ? {
          warn_def: this.props.value.warn_def
        } : null),
        errors: {
          warn_def: this.props.translation.warnings.warn_def
        },
        translation: {
          title: this.props.translation.warnings.title,
          alert: this.props.translation.warnings.alert
        },
        onChange: this.onChangeError
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var nextState = {};

      if (nextProps.value && nextProps.isValid) {
        nextProps.isValid(valid(nextState.value));
      }

      return nextState;
    }
  }]);

  return InputCustom;
}(_react.Component);

exports.InputCustom = InputCustom;
InputCustom.propTypes = {
  /* properties */
  value: _propTypes.default.shape({
    uid: _propTypes.default.string,
    name: _propTypes.default.string,
    required: _propTypes.default.bool,
    is_password: _propTypes.default.bool,
    value: function value(props, propName, componentName) {
      if (props[propName] !== undefined && props["regex"] !== undefined) {
        try {
          var rgx = new RegExp(props[propName]);

          if (!rgx.test(props[propName])) {
            return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "` - invalid value type."));
          }
        } catch (_unused3) {
          return new Error("Invalid prop `regex` - unable to validate value of prop `".concat(propName, "` supplied to `").concat(componentName, "`."));
        }
      }
    },
    default: function _default(props, propName, componentName) {
      if (props[propName] !== undefined && props["regex"] !== undefined) {
        try {
          var rgx = new RegExp(props[propName]);

          if (!rgx.test(props[propName])) {
            return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "` - invalid value type."));
          }
        } catch (_unused4) {
          return new Error("Invalid prop `regex` - unable to validate value of prop `".concat(propName, "` supplied to `").concat(componentName, "`."));
        }
      }
    },
    regex: function regex(props, propName, componentName) {
      if (props[propName] !== undefined) {
        try {
          new RegExp(props[propName]);
        } catch (_unused5) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Invalid regular expression."));
        }
      }
    },
    err_req: _propTypes.default.string,
    err_type: _propTypes.default.string,
    warn_def: _propTypes.default.string,
    placeholder: _propTypes.default.string,
    tip: _propTypes.default.string,
    sm: _propTypes.default.oneOf(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"])
  }),
  translation: _propTypes.default.shape(_propTypes2.CustomTranslationPropType),

  /* functions */
  onChange: _propTypes.default.func,
  isValid: _propTypes.default.func,
  invalid: _propTypes.default.bool,

  /* icons */
  icons: _propTypes.default.shape({
    icon: _propTypes.default.any.isRequired,
    invalid_icon: _propTypes.default.any.isRequired,
    errors: _propTypes.default.any.isRequired,
    errorsAlert: _propTypes.default.any.isRequired
  }),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string
};
InputCustom.defaultProps = {
  className: "",
  translation: _propTypes2.Default_CustomTranslation,
  invalid: false,
  icons: {
    icon: _icons.custom,
    invalid_icon: _icons.invalid_custom,
    errors: _icons.error,
    errorsAlert: _icons.errorAlert,
    warnings: _icons.warnings,
    warningsAlert: _icons.warningAlert
  },
  jkey: _propTypes2.Default_keyPropType,
  size: "sm"
};

var clean = function clean(e) {
  var rgx = true;
  var rgxV = true;
  var rgxD = true;

  try {
    var rg = new RegExp(e.regex);
    rgxV = rg.test(e.value);
    rgxD = rg.test(e.default);
  } catch (_unused6) {
    rgx = false;
    rgxV = false;
    rgxD = false;
  }

  var defValid = e.default.length > 0;
  return Object.assign({}, {
    uid: e.uid
  }, e.name.length > 0 ? {
    name: e.name
  } : null, e.required ? {
    required: e.required
  } : null, e.is_password ? {
    is_password: e.is_password
  } : null, e.value.length > 0 && rgxV ? {
    value: e.value
  } : null, defValid && rgxD && e.required ? {
    default: e.default
  } : null, rgx ? {
    regex: e.regex
  } : null, e.err_req.length > 0 && e.required ? {
    err_req: e.err_req
  } : null, e.err_type.length > 0 ? {
    err_type: e.err_type
  } : null, e.warn_def.length > 0 && defValid && rgxD && e.required ? {
    warn_def: e.warn_def
  } : null, e.placeholder.length > 0 ? {
    placeholder: e.placeholder
  } : null, e.tip.length > 0 ? {
    tip: e.tip
  } : null, {
    type: "rgx"
  }, e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"].includes(e.sm) ? {
    sm: e.sm
  } : null);
};

exports.clean = clean;

var valid = function valid(e) {
  var valid = e.uid.length > 0;

  try {
    var rgx = new RegExp(e.regex);
    valid = valid && (e.default.length === 0 || rgx.test(e.default) || !e.required) && (e.value.length === 0 || rgx.test(e.value));
  } catch (_unused7) {
    valid = false;
  }

  return valid;
};

exports.valid = valid;

var prototype = function prototype() {
  return {
    uid: "",
    name: "",
    required: false,
    is_password: false,
    value: "",
    default: "",
    regex: "",
    err_req: "",
    err_type: "",
    warn_def: "",
    placeholder: "",
    tip: "",
    type: "rgx",
    sm: "12"
  };
};

exports.prototype = prototype;

var rebuild = function rebuild(e) {
  return {
    uid: e.uid !== undefined && e.uid !== null ? String(e.uid) : "",
    name: e.name !== undefined && e.name !== null ? String(e.name) : "",
    required: e.required !== undefined && e.required !== null ? Boolean(e.required) : false,
    is_password: e.is_password !== undefined && e.is_password !== null ? Boolean(e.is_password) : false,
    value: e.value !== undefined && e.value !== null ? String(e.value) : "",
    default: e.default !== undefined && e.default !== null ? String(e.default) : "",
    regex: e.regex !== undefined && e.regex !== null ? String(e.regex) : "",
    err_req: e.err_req !== undefined && e.err_req !== null ? String(e.err_req) : "",
    err_type: e.err_type !== undefined && e.err_type !== null ? String(e.err_type) : "",
    warn_def: e.warn_def !== undefined && e.warn_def !== null ? String(e.warn_def) : "",
    placeholder: e.placeholder !== undefined && e.placeholder !== null ? String(e.placeholder) : "",
    tip: e.tip !== undefined && e.tip !== null ? String(e.tip) : "",
    type: "rgx",
    sm: e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"].includes(e.sm) ? String(e.sm) : "12"
  };
};

exports.rebuild = rebuild;
