// json-obj-form-generator/dist/components/designer/InputListString.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rebuild = exports.prototype = exports.valid = exports.clean = exports.InputListString = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _propTypes2 = require("./propTypes");

var _utils = require("./utils");

var _validators = require("./validators");

var _icons = require("./icons");

var _ErrorContainer = _interopRequireDefault(require("./ErrorContainer"));

var _subcomponents = require("./subcomponents");

var _ListString = require("./subcomponents/ListString");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var InputListString = /*#__PURE__*/function (_Component) {
  _inherits(InputListString, _Component);

  var _super = _createSuper(InputListString);

  function InputListString() {
    var _this;

    _classCallCheck(this, InputListString);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isOpenMain: true,
      invalid: true,
      errPopUp: false,
      uidErrorShow: false,
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "il-str" + _this.props.jkey.sufix)
    });

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.value;
      props.uid = props.uid.replace(" ", "_");

      if (["minLength", "maxLength", "minNo", "maxNo"].includes(e.target.name)) {
        props.value = (0, _ListString.validGetArrays)(props.value, props);
        props.default = (0, _ListString.validGetArrays)(props.default, props);
      }

      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeBool", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.checked;
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeError", function (key, value) {
      var props = this.props.value;
      props[key] = value;
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeList", function (key, value) {
      var props = this.props.value;
      props[key] = value;
      var validation = valid(props);

      if (this.props.onChange) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "collapseMain", function () {
      this.setState({
        isOpenMain: !this.state.isOpenMain
      });
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(InputListString, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var invalid = !valid(this.props.value);
      /* ------ validation ------ */

      var fmax = (0, _validators.filterInt)(this.props.value.maxLength);
      var fmin = (0, _validators.filterInt)(this.props.value.minLength);
      var mnL = (0, _validators.intValid)(this.props.value.minLength);
      var mxL = (0, _validators.intValid)(this.props.value.maxLength);
      var maxValid = this.props.value.maxLength.length === 0 || mxL;
      var minValid = this.props.value.minLength.length === 0 || mnL;
      var rangeValid = true;

      if (mnL && mxL) {
        rangeValid = maxValid && minValid && fmin <= fmax;
      }

      var err_max = fmax > 0 && maxValid && rangeValid;
      var err_min = fmin >= 0 && minValid && rangeValid;
      var fmaxNo = (0, _validators.filterInt)(this.props.value.maxNo);
      var fminNo = (0, _validators.filterInt)(this.props.value.minNo);
      var mnN = (0, _validators.intValid)(this.props.value.minNo);
      var mxN = (0, _validators.intValid)(this.props.value.maxNo);
      var maxNoValid = this.props.value.maxNo.length === 0 || !isNaN(fmaxNo) && isFinite(fmaxNo);
      var minNoValid = this.props.value.minNo.length === 0 || !isNaN(fminNo) && isFinite(fminNo);
      var rangeNoValid = true;

      if (mnN && mxN) {
        rangeNoValid = maxNoValid && minNoValid && fminNo <= fmaxNo;
      }

      var err_maxNo = fmaxNo > 0 && maxNoValid && rangeNoValid;
      var err_minNo = fminNo >= 0 && minNoValid && rangeNoValid;

      var default_section = /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      });

      if (this.props.value.required) {
        default_section = /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: 6
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
          sm: 1,
          size: this.props.size,
          className: "jofgen-D-inputLabel"
        }, this.props.translation.default), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          className: "jofgen-D-input-col"
        }, /*#__PURE__*/_react.default.createElement(_subcomponents.ListString, {
          uid: "default",
          value: this.props.value.default,
          onChange: this.onChangeList,
          minLength: (0, _validators.intValid)(this.props.value.minLength) && rangeValid && fmin >= 0 ? (0, _validators.filterInt)(this.props.value.minLength) : undefined,
          maxLength: (0, _validators.intValid)(this.props.value.maxLength) && rangeValid && fmax >= 0 ? (0, _validators.filterInt)(this.props.value.maxLength) : undefined,
          min: (0, _validators.intValid)(this.props.value.minNo) && rangeNoValid && fminNo >= 0 ? (0, _validators.filterInt)(this.props.value.minNo) : undefined,
          max: (0, _validators.intValid)(this.props.value.maxNo) && rangeNoValid && fmaxNo >= 0 ? (0, _validators.filterInt)(this.props.value.maxNo) : undefined,
          bsSize: this.props.size,
          err_minNo: this.props.translation.messages.v_minNo,
          err_maxNo: this.props.translation.messages.v_maxNo,
          err_maxlength: this.props.translation.messages.v_maxlength,
          err_minlength: this.props.translation.messages.v_minlength,
          none_enabled: true
        }))));
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        className: this.props.className + " jofgen-D-card jofgen-D-string" + (this.props.value.required ? " required" : "") + (invalid || this.props.invalid ? " invalid" : ""),
        style: this.props.style
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardBody, {
        className: "jofgen-D-card-body"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardTitle, null, /*#__PURE__*/_react.default.createElement("table", {
        style: {
          width: "100%"
        }
      }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", {
        className: "jofgen-D-children"
      }, invalid || this.props.invalid ? this.props.icons.invalid_icon : this.props.icons.icon, /*#__PURE__*/_react.default.createElement("b", null, this.props.translation.title)), /*#__PURE__*/_react.default.createElement("td", {
        style: {
          width: "50px",
          textAlign: "right"
        }
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
        size: this.props.size,
        color: "link",
        onClick: this.collapseMain
      }, /*#__PURE__*/_react.default.createElement("svg", {
        className: "jofgen-D-collapse-icon",
        style: {
          transform: this.state.isOpenMain ? "rotate(180deg)" : "rotate(0deg)"
        },
        viewBox: "0 0 24 24"
      }, /*#__PURE__*/_react.default.createElement("path", {
        fill: "#000",
        d: "M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      })))))))), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 2,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.uid), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "uid",
        type: "text",
        value: this.props.value.uid,
        onChange: this.onChange,
        invalid: this.props.value.uid.length === 0 || this.props.invalid,
        bsSize: this.props.size
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.uid), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        check: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        check: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "required",
        type: "checkbox",
        checked: this.props.value.required,
        onChange: this.onChangeBool
      }), " " + this.props.translation.required))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: "6"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 2,
        className: "jofgen-D-inputLabel jofgen-D-inputLabelWithpopUp"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        className: "jofgen-D-col-form-label-sm",
        size: this.props.size
      }, this.props.translations.width), ["1", "2", "3", "4", "5"].includes(this.props.value.sm) ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        id: this.state.gId + "popup",
        style: {
          float: "right"
        },
        onMouseOver: function onMouseOver() {
          _this2.setState({
            alertShow: true
          });
        },
        onMouseOut: function onMouseOut() {
          _this2.setState({
            alertShow: false
          });
        }
      }, this.props.icons_set.alert), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
        target: this.state.gId + "popup",
        isOpen: this.state.alertShow
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.translations.smallWidthAlert))) : null), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CustomInput, {
        name: "sm",
        type: "select",
        value: this.props.value.sm,
        onChange: this.onChange,
        id: this.state.gId + "dropdown",
        bsSize: this.props.size
      }, /*#__PURE__*/_react.default.createElement("option", {
        value: 1
      }, "1 (12 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 2
      }, "2 (6 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 3
      }, "3 (4 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 4
      }, "4 (3 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 5
      }, "5"), /*#__PURE__*/_react.default.createElement("option", {
        value: 6
      }, "6 (2 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 7
      }, "7"), /*#__PURE__*/_react.default.createElement("option", {
        value: 8
      }, "8"), /*#__PURE__*/_react.default.createElement("option", {
        value: 9
      }, "9"), /*#__PURE__*/_react.default.createElement("option", {
        value: 10
      }, "10"), /*#__PURE__*/_react.default.createElement("option", {
        value: 11
      }, "11"), /*#__PURE__*/_react.default.createElement("option", {
        value: 12
      }, "12 (1 ", this.props.translations.columns, ")"))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.name), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "name",
        type: "text",
        value: this.props.value.name,
        onChange: this.onChange,
        bsSize: this.props.size
      })))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.placeholder), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "placeholder",
        type: "text",
        value: this.props.value.placeholder,
        onChange: this.onChange,
        bsSize: this.props.size
      })))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 12
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 1,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.tip), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "tip",
        type: "textarea",
        value: this.props.value.tip,
        onChange: this.onChange,
        bsSize: this.props.size
      }))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Collapse, {
        isOpen: this.state.isOpenMain
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.value), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_subcomponents.ListString, {
        uid: "value",
        value: this.props.value.value,
        onChange: this.onChangeList,
        minLength: (0, _validators.intValid)(this.props.value.minLength) && rangeValid && fmin >= 0 ? (0, _validators.filterInt)(this.props.value.minLength) : undefined,
        maxLength: (0, _validators.intValid)(this.props.value.maxLength) && rangeValid && fmax >= 0 ? (0, _validators.filterInt)(this.props.value.maxLength) : undefined,
        min: (0, _validators.intValid)(this.props.value.minNo) && rangeNoValid && fminNo >= 0 ? (0, _validators.filterInt)(this.props.value.minNo) : undefined,
        max: (0, _validators.intValid)(this.props.value.maxNo) && rangeNoValid && fmaxNo >= 0 ? (0, _validators.filterInt)(this.props.value.maxNo) : undefined,
        bsSize: this.props.size,
        err_minNo: this.props.translation.messages.v_minNo,
        err_maxNo: this.props.translation.messages.v_maxNo,
        err_maxlength: this.props.translation.messages.v_maxlength,
        err_minlength: this.props.translation.messages.v_minlength,
        none_enabled: true
      })))), default_section, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.minLength), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "minLength",
        type: "text",
        step: "1",
        value: this.props.value.minLength,
        onChange: this.onChange,
        min: "0",
        max: this.props.value.maxLength.length > 0 ? this.props.value.maxLength : undefined,
        invalid: !minValid || !rangeValid || fmin < 0,
        bsSize: this.props.size
      }), !rangeValid ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.minLength) : null, fmin < 0 ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.zeroLength) : null, isNaN(fmin) ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.type) : null))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.maxLength), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "maxLength",
        type: "text",
        step: "1",
        value: this.props.value.maxLength,
        onChange: this.onChange,
        min: this.props.value.minLength.length > 0 ? this.props.value.minLength : 0,
        invalid: !maxValid || !rangeValid || fmax < 0,
        bsSize: this.props.size
      }), !rangeValid ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.maxLength) : null, fmax < 0 ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.zeroLength) : null, isNaN(fmax) ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.type) : null))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.minNo), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "minNo",
        type: "text",
        step: "1",
        value: this.props.value.minNo,
        onChange: this.onChange,
        min: "0",
        max: this.props.value.maxNo.length > 0 ? this.props.value.maxNo : undefined,
        invalid: !minNoValid || !rangeNoValid || fminNo < 0,
        bsSize: this.props.size
      }), !rangeNoValid ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.minNo) : null, fminNo < 0 ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.zeroNo) : null, isNaN(fminNo) ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.type) : null))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 1,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.maxNo), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "maxNo",
        type: "text",
        step: "1",
        value: this.props.value.maxNo,
        onChange: this.onChange,
        min: this.props.value.minNo.length > 0 ? this.props.value.minNo : "0",
        invalid: !maxNoValid || !rangeNoValid || fmaxNo < 0,
        bsSize: this.props.size
      }), !rangeNoValid ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.maxNo) : null, fmaxNo < 0 ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.zeroNo) : null, isNaN(fmaxNo) ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.type) : null)))))), /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_ErrorContainer.default, {
        size: this.props.size,
        jkey: {
          prefix: this.state.gId + "-",
          sufix: ""
        },
        icons: {
          icon: this.props.icons.errors,
          invalid_icon: this.props.icons.errorsAlert
        },
        values: Object.assign({}, this.props.value.required ? {
          err_req: this.props.value.err_req
        } : null, err_min ? {
          err_minlength: this.props.value.err_minlength
        } : null, err_max ? {
          err_maxlength: this.props.value.err_maxlength
        } : null, err_minNo ? {
          err_minNo: this.props.value.err_minNo
        } : null, err_maxNo ? {
          err_maxNo: this.props.value.err_maxNo
        } : null),
        errors: {
          err_req: this.props.translation.errors.err_req,
          err_minlength: this.props.translation.errors.err_minlength,
          err_maxlength: this.props.translation.errors.err_maxlength,
          err_minNo: this.props.translation.errors.err_minNo,
          err_maxNo: this.props.translation.errors.err_maxNo
        },
        translation: {
          title: this.props.translation.errors.title,
          alert: this.props.translation.errors.alert
        },
        onChange: this.onChangeError
      })), /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_ErrorContainer.default, {
        size: this.props.size,
        className: "jofgen-D-warning",
        jkey: {
          prefix: this.state.gId + "-",
          sufix: "warn"
        },
        icons: {
          icon: this.props.icons.warnings,
          invalid_icon: this.props.icons.warningsAlert
        },
        values: Object.assign({}, this.props.value.default.values.length > 0 ? {
          warn_def: this.props.value.warn_def
        } : null, {
          empty: this.props.value.empty
        }),
        errors: {
          warn_def: this.props.translation.warnings.warn_def,
          empty: this.props.translation.warnings.empty
        },
        translation: {
          title: this.props.translation.warnings.title,
          alert: this.props.translation.warnings.alert
        },
        onChange: this.onChangeError
      })));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var nextState = {};

      if (nextProps.value && nextProps.isValid) {
        nextProps.isValid(valid(nextState.value));
      }

      return nextState;
    }
  }]);

  return InputListString;
}(_react.Component);

exports.InputListString = InputListString;
InputListString.propTypes = {
  /* properties */
  value: _propTypes.default.shape({
    uid: _propTypes.default.string,
    name: _propTypes.default.string,
    required: _propTypes.default.bool,
    value: _propTypes.default.exact({
      new_value: _propTypes.default.string.isRequired,
      new_min: _propTypes.default.bool.isRequired,
      new_max: _propTypes.default.bool.isRequired,
      values: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,
      emin: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired,
      emax: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired
    }),
    default: _propTypes.default.exact({
      new_value: _propTypes.default.string.isRequired,
      new_min: _propTypes.default.bool.isRequired,
      new_max: _propTypes.default.bool.isRequired,
      values: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,
      emin: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired,
      emax: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired
    }),
    minLength: function minLength(props, propName, componentName) {
      if (props[propName] !== undefined) {
        if (!(0, _validators.intValid)(String(props[propName]))) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be integer."));
        }

        if (props[propName] < 0) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than zero or equal to zero."));
        }

        if (props["maxLength"] !== undefined && props["maxLength"] < props[propName]) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be lower than maxlength."));
        }
      }
    },
    maxLength: function maxLength(props, propName, componentName) {
      if (props[propName] !== undefined) {
        if (!(0, _validators.intValid)(String(props[propName]))) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be integer."));
        }

        if (props[propName] < 0) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than zero or equal to zero."));
        }

        if (props["minLength"] !== undefined && props["minLength"] > props[propName]) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than minlength."));
        }
      }
    },
    minNo: function minNo(props, propName, componentName) {
      if (props[propName] !== undefined) {
        if (!(0, _validators.intValid)(String(props[propName]))) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be integer."));
        }

        if (props[propName] < 0) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than zero or equal to zero."));
        }

        if (props["maxNo"] !== undefined && props["maxNo"] < props[propName]) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be lower than maxNo."));
        }
      }
    },
    maxNo: function maxNo(props, propName, componentName) {
      if (props[propName] !== undefined) {
        if (!(0, _validators.intValid)(String(props[propName]))) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be integer."));
        }

        if (props[propName] < 0) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than zero or equal to zero."));
        }

        if (props["minNo"] !== undefined && props["minNo"] > props[propName]) {
          return new Error("Invalid prop `".concat(propName, "` supplied to `").concat(componentName, "`. Value must be greater than minNo."));
        }
      }
    },
    err_req: _propTypes.default.string,
    empty: _propTypes.default.string,
    err_minlength: _propTypes.default.string,
    err_maxlength: _propTypes.default.string,
    err_minNo: _propTypes.default.string,
    err_maxNo: _propTypes.default.string,
    warn_def: _propTypes.default.string,
    placeholder: _propTypes.default.string,
    tip: _propTypes.default.string,
    sm: _propTypes.default.oneOf(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"])
  }),
  translation: _propTypes.default.shape(_propTypes2.ListStringTranslationPropType),

  /* functions */
  onChange: _propTypes.default.func,
  invalid: _propTypes.default.bool,

  /* icons */
  icons: _propTypes.default.shape({
    icon: _propTypes.default.any.isRequired,
    invalid_icon: _propTypes.default.any.isRequired,
    errors: _propTypes.default.any.isRequired,
    errorsAlert: _propTypes.default.any.isRequired,
    warnings: _propTypes.default.any.isRequired,
    warningsAlert: _propTypes.default.any.isRequired
  }),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string
};
InputListString.defaultProps = {
  className: "",
  translation: _propTypes2.Default_ListStringTranslation,
  invalid: false,
  icons: {
    icon: _icons.inputlist,
    invalid_icon: _icons.invalid_inputlist,
    errors: _icons.error,
    errorsAlert: _icons.errorAlert,
    warnings: _icons.warnings,
    warningsAlert: _icons.warningAlert
  },
  jkey: _propTypes2.Default_keyPropType,
  size: "sm"
};

var clean = function clean(e) {
  var validDefault = !e.default.emin.includes(false) && !e.default.emax.includes(false);
  var validValue = !e.value.emin.includes(false) && !e.value.emax.includes(false);
  var minL = (0, _validators.filterInt)(e.minLength);
  var maxL = (0, _validators.filterInt)(e.maxLength);
  var minNo = (0, _validators.filterInt)(e.minNo);
  var maxNo = (0, _validators.filterInt)(e.maxNo);
  var minValid = (0, _validators.intValid)(e.minNo) && (maxNo >= minNo || isNaN(maxNo)) && minNo >= 0;
  var maxValid = (0, _validators.intValid)(e.maxNo) && (maxNo >= minNo || isNaN(minNo)) && maxNo >= 0;
  var defLen = e.default.values.length > 0;
  var valLen = e.value.values.length > 0;
  var defValid = defLen && (maxNo >= e.default.values.length && maxValid || !maxValid && defLen) && (minNo <= e.default.values.length && minValid || !minValid && defLen) && validDefault;
  var valValid = valLen && (maxNo >= e.value.values.length && maxValid || !maxValid && valLen) && (minNo <= e.value.values.length && minValid || !minValid && valLen) && validValue;
  var minLValid = (0, _validators.intValid)(e.minLength) && (maxL >= minL || isNaN(maxL)) && minL >= 0;
  var maxLValid = (0, _validators.intValid)(e.maxLength) && (maxL >= minL || isNaN(minL)) && maxL >= 0;
  return Object.assign({}, {
    uid: e.uid
  }, e.name.length > 0 ? {
    name: e.name
  } : null, e.required ? {
    required: e.required
  } : null, valValid ? {
    value: e.value.values
  } : null, defValid && e.required ? {
    default: e.default.values
  } : null, minLValid ? {
    minLength: minL
  } : null, maxLValid ? {
    maxLength: maxL
  } : null, minValid ? {
    minNo: minNo
  } : null, maxValid ? {
    maxNo: maxNo
  } : null, e.err_req.length > 0 && e.required ? {
    err_req: e.err_req
  } : null, e.err_minlength.length > 0 && minLValid ? {
    err_minlength: e.err_minlength
  } : null, e.err_maxlength.length > 0 && maxLValid ? {
    err_maxlength: e.err_maxlength
  } : null, e.err_minNo.length > 0 && minValid ? {
    err_minNo: e.err_minNo
  } : null, e.err_maxNo.length > 0 && maxValid ? {
    err_maxNo: e.err_maxNo
  } : null, e.warn_def.length > 0 && defValid && e.required ? {
    warn_def: e.warn_def
  } : null, e.placeholder.length > 0 ? {
    placeholder: e.placeholder
  } : null, e.tip.length > 0 ? {
    tip: e.tip
  } : null, e.empty.length > 0 ? {
    empty: e.empty
  } : null, {
    type: "str_il"
  }, e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"].includes(e.sm) ? {
    sm: e.sm
  } : null);
};

exports.clean = clean;

var valid = function valid(e) {
  var validDefault = !e.default.emin.includes(false) && !e.default.emax.includes(false);
  var validValue = !e.value.emin.includes(false) && !e.value.emax.includes(false);
  var fmin = (0, _validators.filterInt)(e.minNo);
  var fmax = (0, _validators.filterInt)(e.maxNo);
  var fminL = (0, _validators.filterInt)(e.minLength);
  var fmaxL = (0, _validators.filterInt)(e.maxLength);
  var vdl = e.default.values.length;
  var vvl = e.value.values.length;
  var vml = (0, _validators.intValid)(e.minLength);
  var vmxl = (0, _validators.intValid)(e.maxLength);
  var vm = (0, _validators.intValid)(e.minNo);
  var vmx = (0, _validators.intValid)(e.maxNo);
  return e.uid.length > 0 && (e.minLength.length === 0 || vml) && (e.maxLength.length === 0 || vmxl) && (vml && vmxl && fminL <= fmaxL || !vml || !vmxl) && (e.minNo.length === 0 || vm) && (e.maxNo.length === 0 || vmx) && (vm && vmx && fmin <= fmax || !vm || !vmx) && ((vdl <= fmax || !fmax) && validDefault && (vdl >= fmin || !fmin) || vdl === 0 || !e.required) && ((vvl <= fmax || !fmax) && validValue && (vvl >= fmin || !fmin) || vvl === 0);
};

exports.valid = valid;

var prototype = function prototype() {
  return {
    uid: "",
    name: "",
    required: false,
    value: (0, _ListString.prototype)(),
    default: (0, _ListString.prototype)(),
    minLength: "",
    maxLength: "",
    minNo: "",
    maxNo: "",
    err_req: "",
    err_minlength: "",
    err_maxlength: "",
    err_minNo: "",
    err_maxNo: "",
    warn_def: "",
    placeholder: "",
    tip: "",
    empty: "",
    type: "str_il",
    sm: "12"
  };
};

exports.prototype = prototype;

var rebuild = function rebuild(e) {
  return {
    uid: e.uid !== undefined && e.uid !== null ? String(e.uid) : "",
    name: e.name !== undefined && e.name !== null ? String(e.name) : "",
    required: e.required !== undefined && e.required !== null ? Boolean(e.required) : false,
    value: e.value !== undefined && e.value !== null && Array.isArray(e.value) ? (0, _ListString.rebuild)(e.value, e) : (0, _ListString.prototype)(),
    default: e.default !== undefined && e.default !== null && Array.isArray(e.default) ? (0, _ListString.rebuild)(e.default, e) : (0, _ListString.prototype)(),
    minLength: e.minLength !== undefined && e.minLength !== null ? e.minLength : "",
    maxLength: e.maxLength !== undefined && e.maxLength !== null ? e.maxLength : "",
    minNo: e.minNo !== undefined && e.minNo !== null ? e.minNo : "",
    maxNo: e.maxNo !== undefined && e.maxNo !== null ? e.maxNo : "",
    err_req: e.err_req !== undefined && e.err_req !== null ? String(e.err_req) : "",
    err_minlength: e.err_minlength !== undefined && e.err_minlength !== null ? String(e.err_minlength) : "",
    err_maxlength: e.err_maxlength !== undefined && e.err_maxlength !== null ? String(e.err_maxlength) : "",
    err_minNo: e.err_minNo !== undefined && e.err_minNo !== null ? String(e.err_minNo) : "",
    err_maxNo: e.err_maxNo !== undefined && e.err_maxNo !== null ? String(e.err_maxNo) : "",
    warn_def: e.warn_def !== undefined && e.warn_def !== null ? String(e.warn_def) : "",
    placeholder: e.placeholder !== undefined && e.placeholder !== null ? String(e.placeholder) : "",
    tip: e.tip !== undefined && e.tip !== null ? String(e.tip) : "",
    empty: e.empty !== undefined && e.empty !== null ? String(e.empty) : "",
    type: "str_il",
    sm: e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"].includes(e.sm) ? String(e.sm) : "12"
  };
};

exports.rebuild = rebuild;
