// json-obj-form-generator/dist/components/designer/Paragraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rebuild = exports.prototype = exports.valid = exports.clean = exports.Paragraph = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _utils = require("./utils");

var _icons = require("./icons");

var _propTypes2 = require("./propTypes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Paragraph = /*#__PURE__*/function (_Component) {
  _inherits(Paragraph, _Component);

  var _super = _createSuper(Paragraph);

  function Paragraph() {
    var _this;

    _classCallCheck(this, Paragraph);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "p" + _this.props.jkey.sufix),
      warningShow: false
    });

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.value;
      this.setState({
        warningShow: +props.sm <= 4
      });
      var validation = valid(props);

      if (this.props.onChange !== undefined) {
        this.props.onChange(props, validation);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(Paragraph, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var invalid = !valid(this.props.value);
      return /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(this.props.className, "  jofgen-D-card jofgen-D-p ").concat(invalid ? " invalid" : ""),
        style: this.props.style
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardBody, {
        className: "jofgen-D-card-body"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardTitle, null, invalid ? this.props.icons.invalid_icon : this.props.icons.icon, /*#__PURE__*/_react.default.createElement("b", null, this.props.translation.title)), /*#__PURE__*/_react.default.createElement(_reactstrap.Row, null, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: "6"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        sm: 2,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.name), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "title",
        type: "text",
        value: this.props.value.title,
        onChange: this.onChange,
        invalid: invalid,
        bsSize: this.props.size
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.empty)))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: "6"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 2,
        className: "jofgen-D-inputLabel jofgen-D-inputLabelWithpopUp"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        className: "jofgen-D-col-form-label-sm",
        size: this.props.size
      }, this.props.translations.width), ["1", "2", "3", "4", "5"].includes(this.props.value.sm) ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        id: this.state.gId + "popup",
        style: {
          float: "right"
        },
        onMouseOver: function onMouseOver() {
          _this2.setState({
            alertShow: true
          });
        },
        onMouseOut: function onMouseOut() {
          _this2.setState({
            alertShow: false
          });
        }
      }, this.props.icons_set.alert), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
        target: this.state.gId + "popup",
        isOpen: this.state.alertShow
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.translations.smallWidthAlert))) : null), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CustomInput, {
        name: "sm",
        type: "select",
        value: this.props.value.sm,
        onChange: this.onChange,
        id: this.state.gId + "dropdown",
        bsSize: this.props.size
      }, /*#__PURE__*/_react.default.createElement("option", {
        value: 1
      }, "1 (12 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 2
      }, "2 (6 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 3
      }, "3 (4 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 4
      }, "4 (3 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 5
      }, "5"), /*#__PURE__*/_react.default.createElement("option", {
        value: 6
      }, "6 (2 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 7
      }, "7"), /*#__PURE__*/_react.default.createElement("option", {
        value: 8
      }, "8"), /*#__PURE__*/_react.default.createElement("option", {
        value: 9
      }, "9"), /*#__PURE__*/_react.default.createElement("option", {
        value: 10
      }, "10"), /*#__PURE__*/_react.default.createElement("option", {
        value: 11
      }, "11"), /*#__PURE__*/_react.default.createElement("option", {
        value: 12
      }, "12 (1 ", this.props.translations.columns, ")")))))), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        size: this.props.size,
        className: "jofgen-D-inputLabel",
        sm: 1
      }, this.props.translation.text), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "text",
        type: "textarea",
        value: this.props.value.text,
        onChange: this.onChange,
        invalid: invalid,
        bsSize: this.props.size
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.empty)))));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var nextState = {};
      if (nextProps.value) ;
      {
        nextState = {
          warningShow: nextProps.value.sm !== undefined && +nextProps.value.sm <= 4
        };

        if (nextProps.isValid) {
          nextProps.isValid(valid(nextProps.value));
        }
      }
      return nextState;
    }
  }]);

  return Paragraph;
}(_react.Component);

exports.Paragraph = Paragraph;
Paragraph.propTypes = {
  value: _propTypes.default.shape({
    title: _propTypes.default.string,
    text: _propTypes.default.string,
    sm: _propTypes.default.oneOf(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"])
  }).isRequired,

  /* functions */
  onChange: _propTypes.default.func.isRequired,
  isValid: _propTypes.default.func,

  /* icons */
  icons: _propTypes.default.shape({
    icon: _propTypes.default.any.isRequired,
    invalid_icon: _propTypes.default.any.isRequired,
    alert: _propTypes.default.any.isRequired
  }),

  /* aditional */
  translation: _propTypes.default.shape(_propTypes2.ParagraphTranslationPropType),
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string
};
Paragraph.defaultProps = {
  value: {
    title: "",
    text: "",
    sm: "12"
  },
  className: "",

  /* icons */
  icons: {
    icon: _icons.paragraph,
    invalid_icon: _icons.invalid_paragraph,
    alert: _icons.alert
  },

  /* aditional */
  translation: _propTypes2.Default_ParagraphTranslation,
  jkey: _propTypes2.Default_keyPropType,
  size: "sm"
};

var clean = function clean(e) {
  return Object.assign({}, e.title ? {
    title: e.title
  } : null, e.text ? {
    text: e.text
  } : null, {
    type: "p"
  }, e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"].includes(e.sm) ? {
    sm: e.sm
  } : null);
};

exports.clean = clean;

var valid = function valid(e) {
  return (e.title !== undefined && e.title.length > 0 || e.text !== undefined && e.text.length > 0) && e.sm !== undefined && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"].includes(e.sm);
};

exports.valid = valid;

var prototype = function prototype() {
  return {
    title: "",
    text: "",
    type: "p",
    sm: "12"
  };
};

exports.prototype = prototype;

var rebuild = function rebuild(e) {
  return {
    title: e.title !== undefined && e.title !== null ? String(e.title) : "",
    text: e.text !== undefined && e.text !== null ? String(e.text) : "",
    type: "p",
    sm: e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"].includes(e.sm) ? String(e.sm) : "12"
  };
};

exports.rebuild = rebuild;
