// json-obj-form-generator/dist/components/designer/SectionPackage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rebuild = exports.prototype = exports.valid = exports.clean = exports.SectionPackage = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _propTypes2 = require("./propTypes");

var _utils = require("./utils");

var _icons = require("./icons");

var _Container = _interopRequireDefault(require("./Container"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SectionPackage = /*#__PURE__*/function (_Component) {
  _inherits(SectionPackage, _Component);

  var _super = _createSuper(SectionPackage);

  function SectionPackage() {
    var _this;

    _classCallCheck(this, SectionPackage);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isOpenMain: true,
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "secpack" + _this.props.jkey.sufix)
    });

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var props = this.props.value;
      props[e.target.name] = e.target.value;
      props.uid = props.uid.replace(" ", "_");
      var isvalid = valid(props);
      var validation = this.props.valid;
      validation[0] = isvalid;
      var uids = this.props.uids;
      uids[0] = props.uid;
      var sub_items = this.props.designer_aliases[props.type].clean(props);
      sub_items.sub = _toConsumableArray(this.props.sub_items.sub);

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeChildren", function (value, sub_validation, idx) {
      var props = this.props.value;
      props.sub[idx] = value;
      var sub_items = this.props.sub_items;
      sub_items.sub[idx] = this.props.package_designer_aliases[value.type].clean(value);
      var validation = this.props.valid;
      validation[1][idx] = sub_validation;
      var uids = this.props.uids;
      uids[1][idx] = value.uid !== undefined && value.uid !== null ? value.uid : null;

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onChangeSub", function (value, sub_validation, uid, subs, idx) {
      var props = this.props.value;
      var sub_items = this.props.sub_items;
      var uids = this.props.uids;
      var validation = this.props.valid;
      props.sub[idx] = value;
      sub_items.sub[idx] = subs;
      uids[1][idx] = uid;
      validation[1][idx] = sub_validation;

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "addChildren", function (position) {
      var props = this.props.value;
      var validation = this.props.valid;
      var uids = this.props.uids;
      var sub_items = this.props.sub_items;
      props.sub.splice(position, 0, _objectSpread({}, this.props.package_designer_aliases.sec.prototype()));
      sub_items.sub.splice(position, 0, _objectSpread({}, this.props.package_designer_aliases.sec.clean(_objectSpread({}, this.props.designer_aliases.sec.prototype()))));
      uids[1].splice(position, 0, this.props.package_designer_aliases.sec.defaultUid());
      validation[1].splice(position, 0, this.props.package_designer_aliases.sec.defaultValid());

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "removeChildren", function (idx) {
      var props = this.props.value;
      var sub_items = this.props.sub_items;
      var validation = this.props.valid;
      var uids = this.props.uids;
      /* adding new element */

      props.sub.splice(idx, 1);
      sub_items.sub.splice(idx, 1);
      validation[1].splice(idx, 1);
      uids[1].splice(idx, 1);

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "changePosition", function (idx, move) {
      var props = this.props.value;
      var validation = this.props.valid;
      var uids = this.props.uids;
      var sub_items = this.props.sub_items;
      var val_idx_move = props.sub[idx];
      var sub_idx_move = sub_items.sub[idx];
      var uids_idx_move = uids[1][idx];
      var valid_idx_move = validation[1][idx];
      props.sub[idx] = props.sub[idx + move];
      sub_items.sub[idx] = sub_items.sub[idx + move];
      uids[1][idx] = uids[1][idx + move];
      validation[1][idx] = validation[1][idx + move];
      props.sub[idx + move] = val_idx_move;
      sub_items.sub[idx + move] = sub_idx_move;
      uids[1][idx + move] = uids_idx_move;
      validation[1][idx + move] = valid_idx_move;

      if (this.props.onChangeSub) {
        this.props.onChangeSub(props, validation, uids, sub_items);
      }

      if (this.props.isValid) {
        this.props.isValid(validation);
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "collapseMain", function () {
      this.setState({
        isOpenMain: !this.state.isOpenMain
      });
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(SectionPackage, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var invalid = !valid(this.props.value);
      var no_of_elements = this.props.value.sub.length - 1;
      return /*#__PURE__*/_react.default.createElement("div", {
        className: "jofgen-D-card jofgen-D-secpack" + (invalid || this.props.invalid[0] ? " invalid" : ""),
        style: this.props.style
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardBody, {
        className: "jofgen-D-card-body"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CardTitle, null, /*#__PURE__*/_react.default.createElement("table", {
        style: {
          width: "100%"
        }
      }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", {
        className: "jofgen-D-children"
      }, invalid || this.props.invalid[0] ? this.props.icons.invalid_icon : this.props.icons.icon, /*#__PURE__*/_react.default.createElement("b", null, this.props.translation.title)), /*#__PURE__*/_react.default.createElement("td", {
        style: {
          width: "50px",
          textAlign: "right"
        }
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
        size: this.props.size,
        color: "link",
        onClick: this.collapseMain
      }, /*#__PURE__*/_react.default.createElement("svg", {
        className: "jofgen-D-collapse-icon",
        style: {
          transform: this.state.isOpenMain ? "rotate(180deg)" : "rotate(0deg)"
        },
        viewBox: "0 0 24 24"
      }, /*#__PURE__*/_react.default.createElement("path", {
        fill: "#000",
        d: "M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      })))))))), /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 2,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.uid), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "uid",
        type: "text",
        value: this.props.value.uid,
        onChange: this.onChange,
        invalid: this.props.value.uid.length === 0 || this.props.invalid[0],
        bsSize: this.props.size
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
        valid: false
      }, this.props.translation.messages.uid)))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: "6"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true,
        className: "jofgen-D-form-group"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 2,
        className: "jofgen-D-inputLabel jofgen-D-inputLabelWithpopUp"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        className: "jofgen-D-col-form-label-sm",
        size: this.props.size
      }, this.props.translations.width), ["1", "2", "3", "4", "5"].includes(this.props.value.sm) ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        id: this.state.gId + "popup",
        style: {
          float: "right"
        },
        onMouseOver: function onMouseOver() {
          _this2.setState({
            alertShow: true
          });
        },
        onMouseOut: function onMouseOut() {
          _this2.setState({
            alertShow: false
          });
        }
      }, this.props.icons_set.alert), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
        target: this.state.gId + "popup",
        isOpen: this.state.alertShow
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.translations.smallWidthAlert))) : null), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CustomInput, {
        name: "sm",
        type: "select",
        value: this.props.value.sm,
        onChange: this.onChange,
        id: this.state.gId + "dropdown",
        bsSize: this.props.size
      }, /*#__PURE__*/_react.default.createElement("option", {
        value: 1
      }, "1 (12 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 2
      }, "2 (6 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 3
      }, "3 (4 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 4
      }, "4 (3 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 5
      }, "5"), /*#__PURE__*/_react.default.createElement("option", {
        value: 6
      }, "6 (2 ", this.props.translations.columns, ")"), /*#__PURE__*/_react.default.createElement("option", {
        value: 7
      }, "7"), /*#__PURE__*/_react.default.createElement("option", {
        value: 8
      }, "8"), /*#__PURE__*/_react.default.createElement("option", {
        value: 9
      }, "9"), /*#__PURE__*/_react.default.createElement("option", {
        value: 10
      }, "10"), /*#__PURE__*/_react.default.createElement("option", {
        value: 11
      }, "11"), /*#__PURE__*/_react.default.createElement("option", {
        value: 12
      }, "12 (1 ", this.props.translations.columns, ")"))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 6
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 2,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.mode), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        className: "jofgen-D-input-col"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.CustomInput, {
        id: this.state.gId,
        name: "design",
        type: "select",
        className: "dropdown",
        value: this.props.design,
        bsSize: this.props.size,
        onChange: this.onChange
      }, /*#__PURE__*/_react.default.createElement("option", {
        value: "list"
      }, this.props.translation.mode_list), /*#__PURE__*/_react.default.createElement("option", {
        value: "stack"
      }, this.props.translation.mode_stack), /*#__PURE__*/_react.default.createElement("option", {
        value: "unlisted"
      }, this.props.translation.mode_unlisted)))))), /*#__PURE__*/_react.default.createElement(_reactstrap.Collapse, {
        isOpen: this.state.isOpenMain,
        className: "jofgen-D-pack-sub"
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
        sm: 2,
        size: this.props.size,
        className: "jofgen-D-inputLabel"
      }, this.props.translation.sections), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
        sm: 12
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "jofgen-D-ext " + this.props.className,
        style: this.props.style
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Row, {
        style: {
          opacity: "1",
          height: "1.8rem"
        }
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, null, /*#__PURE__*/_react.default.createElement("div", {
        className: "hr"
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
        size: this.props.size,
        color: "light",
        className: "btn-add-element",
        onClick: function onClick(e) {
          _this2.addChildren(0);
        }
      }, this.props.translation.addSection)))), this.props.value.sub.map(function (item, idx) {
        var Tag;

        if (_this2.props.package_designer_aliases[item.type] !== undefined) {
          Tag = _this2.props.package_designer_aliases[item.type].tag;
          return /*#__PURE__*/_react.default.createElement(_react.Fragment, {
            key: idx
          }, /*#__PURE__*/_react.default.createElement(_Container.default, {
            size: _this2.props.size,
            onUpButtonClick: idx > 0 ? function () {
              return _this2.changePosition(idx, -1);
            } : undefined,
            onDownButtonClick: idx < no_of_elements ? function () {
              return _this2.changePosition(idx, 1);
            } : undefined,
            onRemoveButtonClick: function onRemoveButtonClick() {
              return _this2.removeChildren(idx);
            },
            hideDissabled: _this2.props.hideDissabled,
            extras: _this2.props.sub_items.sub[idx],
            icons: _this2.props.icons_set.container,
            extended: _this2.props.extended
          }, /*#__PURE__*/_react.default.createElement(Tag, {
            size: _this2.props.size,
            invalid: _this2.props.invalid[1][idx],
            uids: _this2.props.uids[1][idx],
            valid: _this2.props.valid[1][idx],
            value: item,
            mode: _this2.props.mode,
            sub_items: _this2.props.sub_items.sub[idx],
            className: _this2.props.className,
            jkey: _this2.props.jkey,
            hideDissabled: _this2.props.hideDissabled,
            sm: _this2.props.sm,
            icons_set: _this2.props.icons_set,
            icons: _this2.props.icons_set.types[item.type],
            onChange: function onChange(value, validation) {
              _this2.onChangeChildren(value, validation, idx);
            },
            onChangeSub: function onChangeSub(value, validation, uids, sub_items) {
              _this2.onChangeSub(value, validation, uids, sub_items, idx);
            },
            translation: _this2.props.translations.types[item.type],
            translations: _this2.props.translations,
            designer_aliases: _this2.props.designer_aliases,
            package_designer_aliases: _this2.props.package_designer_aliases,
            extended: _this2.props.extended
          })), /*#__PURE__*/_react.default.createElement("div", {
            className: "jofgen-D-ext " + _this2.props.className,
            style: _this2.props.style
          }, /*#__PURE__*/_react.default.createElement(_reactstrap.Row, {
            style: {
              opacity: "1",
              height: "1.8rem"
            }
          }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, null, /*#__PURE__*/_react.default.createElement("div", {
            className: "hr"
          }), /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
            size: _this2.props.size,
            color: "light",
            className: "btn-add-element",
            onClick: function onClick(e) {
              _this2.addChildren(idx + 1);
            }
          }, _this2.props.translation.addSection)))));
        }

        return null;
      }))))));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var nextState = {};

      if (nextProps.value && nextProps.isValid) {
        nextProps.isValid(valid(nextState.value));
      }

      return nextState;
    }
  }]);

  return SectionPackage;
}(_react.Component);

exports.SectionPackage = SectionPackage;
SectionPackage.propTypes = {
  /* properties */
  value: _propTypes.default.shape({
    uid: _propTypes.default.string,
    design: _propTypes.default.oneOf(["unlisted", "list", "stack"]),
    sub: _propTypes.default.arrayOf(_propTypes.default.object),
    sm: _propTypes.default.oneOf(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"])
  }),
  translation: _propTypes.default.shape(_propTypes2.PackTranslationPropType),

  /* functions */
  onChange: _propTypes.default.func,
  isValid: _propTypes.default.func,
  invalid: _propTypes.default.array,

  /* icons */
  icons: _propTypes.default.shape({
    icon: _propTypes.default.any.isRequired,
    invalid_icon: _propTypes.default.any.isRequired,
    errors: _propTypes.default.any.isRequired,
    errorsAlert: _propTypes.default.any.isRequired
  }),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string,
  uids: _propTypes.default.array
};
SectionPackage.defaultProps = {
  className: "",
  translation: _propTypes2.Default_PackTranslation,
  invalid: [true, []],
  uids: [],
  icons: {
    icon: _icons.section,
    invalid_icon: _icons.invalid_section,
    errors: _icons.error,
    errorsAlert: _icons.errorAlert
  },
  jkey: _propTypes2.Default_keyPropType,
  size: "sm"
};

var clean = function clean(e) {
  return Object.assign({
    uid: e.uid,
    design: e.design,
    type: "pack",
    sub: e.sub
  }, e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"].includes(e.sm) ? {
    sm: e.sm
  } : null);
};

exports.clean = clean;

var valid = function valid(e) {
  return e.uid.length > 0;
};

exports.valid = valid;

var prototype = function prototype() {
  return {
    uid: "",
    design: "list",
    type: "pack",
    sub: [],
    sm: "12"
  };
};

exports.prototype = prototype;

var rebuild = function rebuild(e) {
  return {
    uid: e.uid !== undefined && e.uid !== null ? String(e.uid) : "",
    design: e.mode !== undefined && e.mode !== null && ["unlisted", "list", "stack"].includes(e.mode) ? String(e.mode) : "list",
    type: "pack",
    sub: [],
    sm: e.sm !== undefined && e.sm !== null && ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"].includes(e.sm) ? String(e.sm) : "12"
  };
};

exports.rebuild = rebuild;
