// json-obj-form-generator/dist/components/generator/Container.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _utils = require("./utils");

var _propTypes2 = require("./propTypes");

var _icons = require("./icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Container = /*#__PURE__*/function (_Component) {
  _inherits(Container, _Component);

  var _super = _createSuper(Container);

  function Container() {
    var _this;

    _classCallCheck(this, Container);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "param-" + _this.props.jkey.sufix),
      popUp: false
    });

    _defineProperty(_assertThisInitialized(_this), "showInfoPopUp", function () {
      _this.setState({
        popUp: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "hideInfoPopUp", function () {
      _this.setState({
        popUp: false
      });
    });

    return _this;
  }

  _createClass(Container, [{
    key: "render",
    value: function render() {
      if (this.props.tip !== undefined && this.props.tip.length > 0) {
        return /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: this.props.sm,
          className: this.props.className + " jofgen-item",
          style: this.props.style
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: "6"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
          size: this.props.size,
          style: {
            width: "calc(100% - 22px)"
          }
        }, !this.props.name || this.props.name.length === 0 ? this.props.uid.length > 0 ? this.props.uid : this.state.gId : this.props.name, this.props.required && this.props.req_indicator !== undefined ? /*#__PURE__*/_react.default.createElement("span", {
          className: "jofgen-required-item"
        }, this.props.req_indicator) : null), /*#__PURE__*/_react.default.createElement("span", {
          id: this.state.gId,
          style: {
            float: "right"
          },
          onMouseOver: this.showInfoPopUp,
          onMouseOut: this.hideInfoPopUp
        }, this.props.icon), /*#__PURE__*/_react.default.createElement(_reactstrap.Popover, {
          target: this.state.gId,
          isOpen: this.state.popUp
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.PopoverBody, null, this.props.tip))), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: "6"
        }, this.props.children)));
      } else {
        return /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: this.props.sm,
          className: this.props.className + " jofgen-item jofgen-color",
          style: this.props.style
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Label, {
          sm: 6,
          size: this.props.size,
          className: "jofgen-inputLabel"
        }, !this.props.name || this.props.name.length === 0 ? this.props.uid.length > 0 ? this.props.uid : this.state.gId : this.props.name, this.props.required && this.props.req_indicator !== undefined ? /*#__PURE__*/_react.default.createElement("span", {
          className: "jofgen-required-item"
        }, this.props.req_indicator) : null), /*#__PURE__*/_react.default.createElement(_reactstrap.Col, {
          sm: 6,
          className: "jofgen-input-col"
        }, this.props.children)));
      }
    }
  }]);

  return Container;
}(_react.Component);

exports.default = Container;
Container.propTypes = {
  uid: _propTypes.default.string.isRequired,
  required: _propTypes.default.bool,
  value: _propTypes.default.bool,
  size: _propTypes.default.string,
  onChange: _propTypes.default.func,
  isValid: _propTypes.default.func,
  translation: _propTypes.default.shape(_propTypes2.translationUid),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  icon: _propTypes.default.any
};
Container.defaultProps = {
  uid: "",
  required: false,
  value: false,
  className: "",
  size: "sm",
  jkey: _propTypes2.Default_keyPropType,
  translation: _propTypes2.Default_translationUid,
  icon: _icons.info
};
