// json-obj-form-generator/dist/components/generator/InputColor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getValue = exports.getErrors = exports.InputColor = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _propTypes2 = require("./propTypes");

var _icons = require("./icons");

var _validators = require("./validators");

var _Error = _interopRequireDefault(require("./Error"));

var _Clearer = _interopRequireDefault(require("./Clearer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var InputColor = /*#__PURE__*/function (_Component) {
  _inherits(InputColor, _Component);

  var _super = _createSuper(InputColor);

  function InputColor() {
    var _this;

    _classCallCheck(this, InputColor);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var value = {
        value: e.target.value,
        errors: getErrors(e.target.value, this.props)
      };

      if (this.props.onChange) {
        this.props.onChange(this.props.uid, value, !value.errors.invalid);
      }
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(InputColor, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
        row: true
      }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, null, /*#__PURE__*/_react.default.createElement(_Clearer.default, _extends({}, this.props, {
        clear: this.onChange
      }), /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "value",
        type: "text",
        placeholder: this.props.placeholder,
        value: this.props.value !== null ? this.props.value : "",
        onChange: this.onChange,
        invalid: this.props.invalid || this.props.errors.invalid,
        bsSize: this.props.size
      })), /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
        name: "value",
        type: "color",
        value: this.props.value !== null ? this.props.value : "",
        onChange: this.onChange,
        invalid: this.props.invalid || this.props.errors.invalid,
        bsSize: this.props.size
      }))), /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
        hidden: this.props.errors.warn_def,
        color: "info",
        className: "jofgen-warning jofgen-warning-color"
      }, this.props.warn_def + ' ' + this.props.default, /*#__PURE__*/_react.default.createElement("div", {
        className: "jofgen-color-show",
        style: {
          background: this.props.default
        }
      })), /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
        hidden: this.props.errors.err_req,
        color: "danger",
        className: "jofgen-warning jofgen-warning-color"
      }, this.props.err_req), /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
        hidden: this.props.errors.err_type,
        color: "danger",
        className: "jofgen-warning jofgen-warning-color"
      }, this.props.err_type));
      /*<FormText hidden={!this.props.errors.invalid} color="danger" className="jofgen-warning jofgen-warning-color">{this.props.translation.invalidUid}</FormText>*/
    }
  }]);

  return InputColor;
}(_react.Component);

exports.InputColor = InputColor;
InputColor.propTypes = {
  uid: _propTypes.default.string,
  name: _propTypes.default.string,
  required: _propTypes.default.bool,
  value: function value(props, propName, componentName) {
    if (props[propName] !== undefined && !(0, _validators.isRGB)(props[propName])) {
      return new _Error.default('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. RGB color validation failed.');
    }
  },
  default: function _default(props, propName, componentName) {
    if (props[propName] !== undefined && !(0, _validators.isRGB)(props[propName])) {
      return new _Error.default('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. RGB color validation failed.');
    }
  },
  tip: _propTypes.default.string,
  placeholder: _propTypes.default.string,
  errors: _propTypes.default.object,
  err_req: _propTypes.default.string,
  err_type: _propTypes.default.string,
  warn_def: _propTypes.default.string,
  sm: _propTypes.default.oneOf(["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"]),
  size: _propTypes.default.string,
  onChange: _propTypes.default.func,
  isValid: _propTypes.default.func,
  invalid: _propTypes.default.bool,
  translation: _propTypes.default.shape(_propTypes2.translationUid),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  icon: _propTypes.default.any
};
InputColor.defaultProps = {
  uid: "",
  className: "",
  invalid: false,
  value: null,
  err_req: "Value is mandatory",
  err_type: "Invalid value",
  warn_def: "Value is not filled. Default value:",
  placeholder: "#FFFFFF",
  translation: _propTypes2.Default_translationUid,
  size: "sm",
  jkey: _propTypes2.Default_keyPropType,
  sm: "12",
  icon: _icons.info
};

var getErrors = function getErrors(e, props) {
  var new_value = getValue(e, props);

  if (new_value !== undefined && new_value !== null) {
    var w_d = !(e.length === 0 && props.default !== undefined && props.default !== null && props.required);
    var e_r = !(new_value.length === 0 && props.required);
    var e_t = (0, _validators.isRGB)(new_value) || new_value.length === 0 && props.required;
    return {
      err_req: e_r,
      warn_def: w_d,
      err_type: e_t,
      invalid: !(e_r && e_t)
    };
  } else {
    return {
      err_req: true,
      warn_def: true,
      err_type: true,
      invalid: false
    };
  }
};

exports.getErrors = getErrors;

var getValue = function getValue(e, props) {
  if (props.required) {
    if (e === undefined || e === null || e.length === 0) {
      if (props.default === null || props.default === undefined) {
        return "";
      }

      return props.default;
    } else {
      return e;
    }
  } else {
    if (e === undefined || e === null) {
      return null;
    } else {
      return e;
    }
  }
};

exports.getValue = getValue;
