// json-obj-form-generator/dist/components/generator/propTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Default_translationUid = exports.translationUid = exports.Default_keyPropType = exports.keyPropType = exports.InputWithPopUpPropType = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InputWithPopUpPropType = {
  name: _propTypes.default.string.isRequired,
  tip: _propTypes.default.string.isRequired,
  placeHolder: _propTypes.default.string.isRequired
};
exports.InputWithPopUpPropType = InputWithPopUpPropType;
var keyPropType = {
  prefix: _propTypes.default.string.isRequired,
  sufix: _propTypes.default.string.isRequired
};
exports.keyPropType = keyPropType;
var Default_keyPropType = {
  prefix: "jofgen-",
  sufix: ""
};
exports.Default_keyPropType = Default_keyPropType;
var translationUid = {
  invalidUid: _propTypes.default.string.isRequired,
  noUid: _propTypes.default.string.isRequired
};
exports.translationUid = translationUid;
var Default_translationUid = {
  invalidUid: "Id is invalid or not unique.",
  noUid: "Unique id is not defined. Generic id will be used."
};
exports.Default_translationUid = Default_translationUid;
