// json-obj-form-generator/dist/components/shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.package_designer_aliases = exports.designer_aliases = exports.generator_aliases = void 0;

var _InputBoolean = require("./designer/InputBoolean");

var _InputColor = require("./designer/InputColor");

var _InputCustom = require("./designer/InputCustom");

var _InputFloat = require("./designer/InputFloat");

var _InputInteger = require("./designer/InputInteger");

var _InputString = require("./designer/InputString");

var _InputListCustom = require("./designer/InputListCustom");

var _InputListFloat = require("./designer/InputListFloat");

var _InputListInteger = require("./designer/InputListInteger");

var _InputListString = require("./designer/InputListString");

var _Paragraph = require("./designer/Paragraph");

var _Section = require("./designer/Section");

var _SectionPackage = require("./designer/SectionPackage");

var _generator = require("./generator");

var generator_aliases = {
  //primitives
  bool: {
    tag: _generator.InputBoolean,
    useContainer: true
  },
  color: {
    tag: _generator.InputColor,
    useContainer: true
  },
  rgx: {
    tag: _generator.InputCustom,
    useContainer: true
  },
  float: {
    tag: _generator.InputFloat,
    useContainer: true
  },
  int: {
    tag: _generator.InputInteger,
    useContainer: true
  },
  p: {
    tag: _generator.Paragraph,
    useContainer: false
  },
  str: {
    tag: _generator.InputString,
    useContainer: true
  },
  // input lists
  float_il: {
    tag: _generator.InputListFloat,
    useContainer: true
  },
  int_il: {
    tag: _generator.InputListInteger,
    useContainer: true
  },
  rgx_il: {
    tag: _generator.InputListCustom,
    useContainer: true
  },
  str_il: {
    tag: _generator.InputListString,
    useContainer: true
  },
  sec: {
    tag: _generator.Section,
    useContainer: false
  },
  pack: {
    tag: _generator.SectionPackage,
    useContainer: false
  },
  err: {
    tag: _generator.Error,
    useContainer: false
  }
};
exports.generator_aliases = generator_aliases;
var designer_aliases = {
  //primitives
  bool: {
    value: false,
    tag: _InputBoolean.InputBoolean,
    clean: _InputBoolean.clean,
    prototype: _InputBoolean.prototype,
    valid: _InputBoolean.valid,
    rebuild: _InputBoolean.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  color: {
    value: '',
    tag: _InputColor.InputColor,
    clean: _InputColor.clean,
    prototype: _InputColor.prototype,
    valid: _InputColor.valid,
    rebuild: _InputColor.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  rgx: {
    value: '',
    tag: _InputCustom.InputCustom,
    clean: _InputCustom.clean,
    prototype: _InputCustom.prototype,
    valid: _InputCustom.valid,
    rebuild: _InputCustom.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  float: {
    value: '',
    tag: _InputFloat.InputFloat,
    clean: _InputFloat.clean,
    prototype: _InputFloat.prototype,
    valid: _InputFloat.valid,
    rebuild: _InputFloat.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  int: {
    value: '',
    tag: _InputInteger.InputInteger,
    clean: _InputInteger.clean,
    prototype: _InputInteger.prototype,
    valid: _InputInteger.valid,
    rebuild: _InputInteger.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  p: {
    value: null,
    tag: _Paragraph.Paragraph,
    clean: _Paragraph.clean,
    prototype: _Paragraph.prototype,
    valid: _Paragraph.valid,
    rebuild: _Paragraph.rebuild,
    mustHaveUid: false,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return null;
    }
  },
  str: {
    value: '',
    tag: _InputString.InputString,
    clean: _InputString.clean,
    prototype: _InputString.prototype,
    valid: _InputString.valid,
    rebuild: _InputString.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  // input lists
  float_il: {
    value: [],
    tag: _InputListFloat.InputListFloat,
    clean: _InputListFloat.clean,
    prototype: _InputListFloat.prototype,
    valid: _InputListFloat.valid,
    rebuild: _InputListFloat.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  int_il: {
    value: [],
    tag: _InputListInteger.InputListInteger,
    clean: _InputListInteger.clean,
    prototype: _InputListInteger.prototype,
    valid: _InputListInteger.valid,
    rebuild: _InputListInteger.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  rgx_il: {
    value: [],
    tag: _InputListCustom.InputListCustom,
    clean: _InputListCustom.clean,
    prototype: _InputListCustom.prototype,
    valid: _InputListCustom.valid,
    rebuild: _InputListCustom.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  str_il: {
    value: [],
    tag: _InputListString.InputListString,
    clean: _InputListString.clean,
    prototype: _InputListString.prototype,
    valid: _InputListString.valid,
    rebuild: _InputListString.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return false;
    },
    defaultUid: function defaultUid() {
      return '';
    }
  },
  // sections
  sec: {
    value: null,
    tag: _Section.Section,
    clean: _Section.clean,
    prototype: _Section.prototype,
    valid: _Section.valid,
    rebuild: _Section.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return [false, []];
    },
    defaultUid: function defaultUid() {
      return ['', []];
    }
  },
  pack: {
    value: null,
    tag: _SectionPackage.SectionPackage,
    clean: _SectionPackage.clean,
    prototype: _SectionPackage.prototype,
    valid: _SectionPackage.valid,
    rebuild: _SectionPackage.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return [false, []];
    },
    defaultUid: function defaultUid() {
      return ['', []];
    }
  }
};
exports.designer_aliases = designer_aliases;
var package_designer_aliases = {
  sec: {
    value: null,
    tag: _Section.Section,
    clean: _Section.clean,
    prototype: _Section.prototype,
    valid: _Section.valid,
    rebuild: _Section.rebuild,
    mustHaveUid: true,
    defaultValid: function defaultValid() {
      return [false, []];
    },
    defaultUid: function defaultUid() {
      return ['', []];
    }
  }
};
exports.package_designer_aliases = package_designer_aliases;
