// json-obj-form-generator/dist/components/generator/subcomponents/ListNumeric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prototype = exports.validGetArrays = exports.ListNumeric = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactstrap = require("reactstrap");

var _icons = require("../icons");

var _propTypes2 = require("../propTypes");

var _reactTransitionGroup = require("react-transition-group");

var _validators = require("../validators");

var _utils = require("./../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ListNumeric = /*#__PURE__*/function (_Component) {
  _inherits(ListNumeric, _Component);

  var _super = _createSuper(ListNumeric);

  function ListNumeric() {
    var _this;

    _classCallCheck(this, ListNumeric);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      gId: (0, _utils.idGenerator)(_this.props.jkey.prefix + "-lint-" + _this.props.jkey.sufix)
    });

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var props = this.props.value;
      var validation_result = isValid(e.target.value, this.props);

      switch (e.target.name) {
        case "value-new":
          props.new_value = e.target.value;
          props.new_min = props.new_value.length === 0 || validation_result[0];
          props.new_max = props.new_value.length === 0 || validation_result[1];
          props.new_type = props.new_value.length === 0 || validation_result[2];

          if (props.new_type && props.new_value.length > 0) {
            props.new_value = this.props.typeFilter(props.new_value);
          }

          break;

        default:
          var idx = parseInt(e.target.name.replace("value-", ""));
          props.emin[idx] = validation_result[0];
          props.emax[idx] = validation_result[1];
          props.etype[idx] = validation_result[2];

          if (validation_result[2]) {
            props.values[idx] = this.props.typeFilter(e.target.value);
          } else {
            props.values[idx] = e.target.value;
          }

      }

      if (this.props.onChange !== undefined) {
        this.props.onChange(this.props.uid, props, !props.emin.includes(false) && !props.emax.includes(false) && !props.etype.includes(false));
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onAdd", function () {
      var props = this.props.value;
      var validation_result = isValid(props.new_value, this.props);
      props.values.push(props.new_value);
      props.emin.push(validation_result[0]);
      props.emax.push(validation_result[1]);
      props.etype.push(validation_result[2]);
      props.new_value = "";
      props.new_min = true;
      props.new_max = true;
      props.new_type = true;

      if (this.props.onChange !== undefined) {
        this.props.onChange(this.props.uid, props, !props.emax.includes(false) && !props.emin.includes(false) && !props.etype.includes(false));
      }
    }.bind(_assertThisInitialized(_this)));

    _defineProperty(_assertThisInitialized(_this), "onDrop", function (e) {
      var props = this.props.value;
      var idx = parseInt(e.currentTarget.value.replace("btn-", ""));
      props.values.splice(idx, 1);
      props.emin.splice(idx, 1);
      props.emax.splice(idx, 1);
      props.etype.splice(idx, 1);

      if (this.props.onChange !== undefined) {
        this.props.onChange(this.props.uid, props, !props.emax.includes(false) && !props.emin.includes(false));
      }
    }.bind(_assertThisInitialized(_this)));

    return _this;
  }

  _createClass(ListNumeric, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var new_element;
      var warn_mode;

      if (this.props.value.values.length < this.props.maxNo) {
        new_element = /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          key: this.state.gId + "-new",
          row: true,
          className: "jofgen-card-lst-row"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, null, /*#__PURE__*/_react.default.createElement("table", {
          className: "jofgen-card-lst-table"
        }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
          name: "value-new",
          type: "text",
          bsSize: this.props.size,
          value: this.props.value.new_value,
          invalid: !this.props.value.new_min || !this.props.value.new_max || !this.props.value.new_type,
          onChange: this.onChange
        }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: this.props.value.new_type,
          color: "danger"
        }, this.props.err_type), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: this.props.value.new_min,
          color: "danger"
        }, this.props.err_min), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: this.props.value.new_max,
          color: "danger"
        }, this.props.err_max)), /*#__PURE__*/_react.default.createElement("td", {
          className: "btns"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
          className: "btn-add",
          color: "light",
          size: "sm",
          onClick: this.onAdd
        }, this.props.icons.add))))))));
      }

      if (!this.props.none_enabled || this.props.value.values.length > 0) {
        warn_mode = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, this.props.minNo !== undefined && this.props.value.values.length < this.props.minNo ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
          color: "danger"
        }, this.props.err_minNo) : null, this.props.maxNo !== undefined && this.props.value.values.length > this.props.maxNo ? /*#__PURE__*/_react.default.createElement(_reactstrap.FormText, {
          color: "danger"
        }, this.props.err_maxNo) : null);
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        className: this.props.className + "jofgen-card-lst",
        style: this.props.style
      }, new_element, warn_mode, /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.TransitionGroup, null, this.props.value.values.map(function (item, idx) {
        return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.CSSTransition, {
          key: idx,
          classNames: "jofgen-card-lst-row-item",
          timeout: 50
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.FormGroup, {
          row: true,
          className: "jofgen-card-lst-row"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Col, null, /*#__PURE__*/_react.default.createElement("table", {
          className: "jofgen-card-lst-table"
        }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_reactstrap.Input, {
          name: "value-" + idx,
          type: "text",
          bsSize: _this2.props.size,
          value: item,
          invalid: !_this2.props.value.emin[idx] || !_this2.props.value.emax[idx] || !_this2.props.value.etype[idx],
          onChange: _this2.onChange
        }), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: _this2.props.value.etype[idx],
          color: "danger"
        }, _this2.props.err_type), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: _this2.props.value.emin[idx],
          color: "danger"
        }, _this2.props.err_min), /*#__PURE__*/_react.default.createElement(_reactstrap.FormFeedback, {
          hidden: _this2.props.value.emax[idx],
          color: "danger"
        }, _this2.props.err_max)), /*#__PURE__*/_react.default.createElement("td", {
          className: "btns"
        }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
          className: "btn-remove",
          value: "btn-" + idx,
          color: "light",
          size: "sm",
          onClick: _this2.onDrop
        }, _this2.props.icons.drop))))))));
      })));
    }
  }]);

  return ListNumeric;
}(_react.Component);

exports.ListNumeric = ListNumeric;
ListNumeric.propTypes = {
  /* properties */
  uid: _propTypes.default.string.isRequired,
  value: _propTypes.default.exact({
    new_value: _propTypes.default.string.isRequired,
    new_min: _propTypes.default.bool.isRequired,
    new_max: _propTypes.default.bool.isRequired,
    new_type: _propTypes.default.bool.isRequired,
    values: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])).isRequired,
    emin: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired,
    emax: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired,
    etype: _propTypes.default.arrayOf(_propTypes.default.bool).isRequired
  }),
  min: _propTypes.default.number,
  max: _propTypes.default.number,
  minNo: function minNo(props, propName, componentName) {
    if (props[propName] !== undefined) {
      if (!(0, _validators.intValid)(String(props[propName]))) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be integer.');
      }

      if (props[propName] < 0) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be greater than zero or equal to zero.');
      }

      if (props["maxNo"] !== undefined && props["maxNo"] < props[propName]) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be lower than maxNo.');
      }
    }
  },
  maxNo: function maxNo(props, propName, componentName) {
    if (props[propName] !== undefined) {
      if (!(0, _validators.intValid)(String(props[propName]))) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be integer.');
      }

      if (props[propName] < 0) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be greater than zero or equal to zero.');
      }

      if (props["minNo"] !== undefined && props["minNo"] > props[propName]) {
        return new Error('Invalid prop `' + propName + '` supplied to' + ' `' + componentName + '`. Value must be greater than minNo.');
      }
    }
  },
  err_min: _propTypes.default.string,
  err_max: _propTypes.default.string,
  err_minNo: _propTypes.default.string,
  err_maxNo: _propTypes.default.string,
  err_type: _propTypes.default.string,
  none_enabled: _propTypes.default.bool,

  /* functions */
  onChange: _propTypes.default.func,
  invalid: _propTypes.default.bool,

  /* icons */
  icons: _propTypes.default.shape({
    add: _propTypes.default.any.isRequired,
    drop: _propTypes.default.any.isRequired
  }),

  /* aditional */
  jkey: _propTypes.default.shape(_propTypes2.keyPropType),
  size: _propTypes.default.string
};
ListNumeric.defaultProps = {
  className: "",
  invalid: false,
  min: -Infinity,
  max: Infinity,
  minNo: 0,
  maxNo: Infinity,
  err_minNo: "Number of values is lower than minimum",
  err_maxNo: "Number of values is higher than maximum",
  err_max: "Value is too high",
  err_min: "Value is too low",
  err_type: "Invalid value type. Value must be integer",
  icons: {
    add: _icons.plus,
    drop: _icons.trashCan
  },
  jkey: _propTypes2.Default_keyPropType,
  size: "sm",
  none_enabled: false
};

var isValid = function isValid(value, props) {
  var valid = [];
  var v = String(value);
  var val = props.typeFilter(v);
  var valValid = props.typeValid(v);

  if (props.min !== undefined && valValid) {
    valid.push(val >= props.min);
  } else {
    valid.push(true);
  }

  if (props.max !== undefined && valValid) {
    valid.push(val <= props.max);
  } else {
    valid.push(true);
  }

  valid.push(valValid);
  return valid;
};

var validGetArrays = function validGetArrays(e, props, typeValid, typeFilter) {
  var snv = String(e.new_value);
  var result = {
    new_value: "",
    new_min: true,
    new_max: true,
    new_type: typeValid(snv),
    values: [],
    emin: [],
    emax: [],
    etype: []
  };
  var new_value_empty = snv.length === 0;
  result.new_value = result.new_type ? typeFilter(snv) : snv;
  var values = [];
  var emin = [];
  var emax = [];
  var etype = [];
  var min = props.min === undefined || !typeValid(props.min) ? -Infinity : typeFilter(props.min);
  var max = props.max === undefined || !typeValid(props.max) ? Infinity : typeFilter(props.max);

  if (result.new_type) {
    result.new_min = min <= result.new_value || new_value_empty;
    result.new_max = max >= result.new_value || new_value_empty;
  }

  result.new_type = result.new_type || new_value_empty;

  if (e.values !== undefined) {
    values = e.values;
    var valuesLength = values.length;
    var svi;
    var ivi;
    var vvi;

    for (var i = 0; i < valuesLength; i++) {
      svi = String(values[i]);
      ivi = typeFilter(svi);
      vvi = typeValid(svi);

      if (vvi) {
        emin.push(ivi >= min);
        emax.push(ivi <= max);
        etype.push(vvi);
      } else {
        emin.push(true);
        emax.push(true);
        etype.push(vvi);
      }

      result.values.push(vvi ? ivi : svi);
    }
  }

  result.emin = emin;
  result.emax = emax;
  result.etype = etype;
  return result;
};

exports.validGetArrays = validGetArrays;

var prototype = function prototype() {
  return {
    new_value: "",
    new_min: true,
    new_max: true,
    new_type: true,
    values: [],
    emin: [],
    emax: [],
    etype: []
  };
};

exports.prototype = prototype;
